aspnet/fundamentals/logging.rst

```diff
@@ -126,105 +180,251 @@ Debug          dbug
 
-In addition to the built-in loggers, you can configure logging to use other providers. Add the appropriate package to your *project.json* file, and then configure it just like any other provider. Typically, these packages include extension methods on ``ILoggerFactory`` to make it easy to add them.
+To use ``TraceSource``, add the ``Microsoft.Extensions.Logging.TraceSource`` package to your project, along with any specific trace source packages you'll be using, such as ``TextWriterTraceListener`` in the sample application that goes with this article.
+
+The following example demonstrates how to configure a ``TraceSourceLogger`` instance for an application, with a listener that processes only ``Warning`` or higher priority messages. Each call to ``AddTraceSource`` takes a ``TraceListener``. The code shown configures a ``TextWriterTraceListener`` to write to the console window.
+
+.. literalinclude:: logging/sample/src/TodoApi/Startup.cs
+  :language: c#
+  :start-after: snippet_TraceSource
+  :end-before: endregion
+  :emphasize-lines: 8-12
+  :dedent: 8
 
- * `elmah.io <https://github.com/elmahio/Elmah.Io.Extensions.Logging>`_ - provider for the elmah.io service
+The ``sourceSwitch`` is configured to use ``SourceLevels.Warning``, so only ``Warning`` (or higher) log messages are picked up by the ``TraceListener`` instance.
+
+To see a log displayed by the ``TextWriterTraceListener``, run the app from the command line and navigate to ``http://localhost:5000/api/Todo/0``. You see output similar to the following:
+
+.. code-block:: none
+
+  Now listening on: http://localhost:5000
+  Application started. Press Ctrl+C to shut down.
+  TodoApi.Controllers.TodoController Warning: 4000 : GetById(0) NOT FOUND
+
+There are many other ``TraceSource`` listeners available, and the ``TextWriterTraceListener`` can be configured to use any ``TextWriter`` instance, making this a very flexible option for logging.
+
+Third-party logging frameworks
+------------------------------
+
+Here are some third-party logging frameworks that work with ASP.NET Core: 
+
+ * `elmah.io <https://github.com/elmahio/Elmah.Io.Extensions.Logging>`_ - provider for the Elmah.Io service
  * `Loggr <https://github.com/imobile3/Loggr.Extensions.Logging>`_ - provider for the Loggr service
  * `NLog <https://github.com/NLog/NLog.Extensions.Logging>`_ - provider for the NLog library
  * `Serilog <https://github.com/serilog/serilog-framework-logging>`_ - provider for the Serilog library
 
+Some third-party frameworks can do `semantic logging, also known as structured logging <http://programmers.stackexchange.com/questions/312197/benefits-of-structured-logging-vs-basic-logging>`__.  
+
+To use a third-party framework, the process is typically similar to what you've seen for built-in providers: add the appropriate NuGet package to your project and call an extension method on ``ILoggerFactory``. For more information, see each framework's documentation.  
```

Comment:
The third-party frameworks are generally configured similar to the built in providers: